DigiMeSDK/Core/Classes/Security/DMEJWTUtility.swift

```diff
@@ -0,0 +1,446 @@
+        claims.nonce = (DMECryptoUtilities.randomBytes(withLength: 16) as NSData).hexString()
+        claims.codeChallengeMethod = "S256"
+        
+        // NB! this redirect schema must exist in the CA contract definition, otherwise preauth request will fail!
+        claims.redirectUrl = "digime-ca-\(appId)"
+        claims.responseMode = "query"
+        claims.responseType = "code"
+        claims.state = (DMECryptoUtilities.randomBytes(withLength: 32) as NSData).hexString()
+        claims.timestamp = NSDate().timeIntervalSince1970 * 1000.0
+
+        // signing
+        var jwt = JWT(header: header, claims: claims)
+        let signer = JWTSigner.ps512(privateKey: privateKeyData)
+        guard let signedJwt = try? jwt.sign(using: signer) else {
+            print("DigiMeSDK: Error signing preAuth JWT")
+            return nil
+        }
+
+        // validation
+        guard
+            let publicKeyBase64 = publicKey,
+            let publicKey = convertKeyString(publicKeyBase64) else {
+                return signedJwt
+        }
+        
+        let verifier = JWTVerifier.ps512(publicKey: publicKey)
+        let isVerified = JWT<PayloadRequestPreauthJWT>.verify(signedJwt, using: verifier)
+
+        return isVerified ? signedJwt : nil
+    }
+    
+    /// Creates request JWT which can be used to get an authentication token
+    /// - Parameters:
+    ///   - authCode: OAuth authorization code
+    ///   - appId: application identifier
+    ///   - contractId: contract identifier
+    ///   - privateKey: private key in base 64 format
+    ///   - publicKey: public key in base 64 format
+    @objc public class func signedAuthJwt(_ authCode: String, appId: String, contractId: String, privateKey: String, publicKey: String?) -> String? {
+        guard
+            privateKey.isHex(),
+            let privateKeyData = convertKeyString(privateKey) else {
+                print("DigiMeSDK: Error creating RSA key")
+                return nil
+        }
+
+        let claims = PayloadRequestAuthJWT()
+        claims.clientId = "\(appId)_\(contractId)"
+        claims.code = authCode
+        claims.codeVerifier = retrieveCodeVerifier()
+        claims.grantType = "authorization_code"
+        claims.nonce = (DMECryptoUtilities.randomBytes(withLength: 16) as NSData).hexString()
+        
+        // NB! this redirect schema must exist in the CA contract definition, otherwise preauth request will fail!
+        claims.redirectUrl = "digime-ca-\(appId)"
+        claims.timestamp = NSDate().timeIntervalSince1970 * 1000.0
+
+        // signing
+        var jwt = JWT(header: header, claims: claims)
+        let signer = JWTSigner.ps512(privateKey: privateKeyData)
+        guard let signedJwt = try? jwt.sign(using: signer) else {
+            print("DigiMeSDK: Error signing auth JWT")
+            return nil
+        }
+
+        // validation
+        guard
+            let publicKeyBase64 = publicKey,
```

Comment:
Hmm, but the public key **IS** expected in `Base64`. While parameters can be of different types - of course - given that thes are keys, we may as well keep the key format consistent. I smell refactoring.... somewhere.... 😄